Sources/SwiftSyntax/SyntaxComparison.swift

```diff
@@ -0,0 +1,142 @@
+//===--------- SyntaxComparison.swift - Syntax Comparison -----------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Utility APIs to compare two difference syntax trees for structual
+// equivalence.
+//
+//===----------------------------------------------------------------------===//
+
+public enum DifferenceReason {
+  case nodeType, presence, missingNode, additionalNode, trivia, tokenKind, tokenText
+}
+
+public struct TreeDifference {
+  /// The `node` that's different to `baseline`, unless `reason` is
+  /// `.missingNode`. In that case it's actually the parent that's missing a
+  /// child node.
+  public let node: Syntax
+  /// The corresponding `baseline` that does not match `node`, unless `reason`
+  /// is `.additionalNode`. In that case it's what would be the parent (if the
+  /// baseline node existed).
+  public let baseline: Syntax
+  public let reason: DifferenceReason
+
+  public init(node: Syntax, baseline: Syntax, reason: DifferenceReason) {
+    self.node = node
+    self.baseline = baseline
+    self.reason = reason
+  }
+}
+
+extension TreeDifference: CustomDebugStringConvertible {
+  public var debugDescription: String {
+    let verbose = reason == .trivia
+    let expectedDesc = baseline.debugDescription(includeChildren: false, includeTrivia: verbose)
+    let actualDesc = node.debugDescription(includeChildren: false, includeTrivia: verbose)
+
+    let message: String
+    switch reason {
+    case .nodeType:
+      fallthrough
+    case .presence:
+      fallthrough
+    case .trivia:
+      fallthrough
+    case .tokenKind:
+      fallthrough
+    case .tokenText:
```

Comment:
I think we can rewrite this to a slightly more-concise multi-case

```suggestion
    case .nodeType, .presence, .trivia, .tokenKind, .tokenText:
```

Reply:
Oof yeah, thanks. I clearly have too much C-language-style in me 😆.